File: binaryrightview.java
/*
Time Complexity: O(N)
Space Complexity: O(H) where h is the height of the tree
DFS we are overwriting every time moving from left to right on each level
*/
class Solution {
    void helper(TreeNode current,int level, List<Integer> result){    
        /*initiaizing for the first time*/
        if(result.size()<=level){
        result.add(level,current.val);
        }
        /*replacing the second time*/
        else{
            result.set(level,current.val);
        }
        if(current.left!=null){
         helper(current.left,level+1,result);   
        }
        if(current.right!=null){
        helper(current.right,level+1,result);   
        }
    }
    public List<Integer> rightSideView(TreeNode root) {
        List<Integer> result=new ArrayList<Integer>();
        if(root==null){
            return result;
        }
        helper(root,0,result);
        return result;
    }
}

File: BinaryRightView.cpp
//time complexity: O(n)
//space complexity: O(n)

//pPerformed BFS and took the last element for each level
class Solution {
public:
    vector<int> rightSideView(TreeNode* root) {
        if(root==NULL) return {};
        queue<TreeNode*> q;
        q.push(root);
        vector<int> li;



        while(!q.empty()){
            int size=q.size();
            
            for(int i=0;i<size;i++){
                TreeNode* curr = q.front();
                q.pop();

                if(i==size-1)li.push_back(curr->val);

                if(curr->left!=NULL){
                    q.push(curr->left);
                }
                if(curr->right!=NULL){
                    q.push(curr->right);
                }


            }
        
            
            


        }

        return li;
        
    }
};